include/emp/prefab/ControlPanel.hpp

```diff
@@ -0,0 +1,331 @@
+        {"MILLISECONDS", 100}, {"FRAMES", 5}
+      };
+
+      // Map units to cool refresh checkers
+      const std::map<std::string, const checker_func_t> refresh_checkers{
+        { "MILLISECONDS",
+          [elapsed_milliseconds = 0, this]
+          (const web::Animate & anim) mutable {
+            int rate = this->refresh_rates[this->refresh_unit];
+            elapsed_milliseconds += anim.GetStepTime();
+            if (elapsed_milliseconds > rate) {
+              elapsed_milliseconds -= rate;
+              if (elapsed_milliseconds > rate) elapsed_milliseconds = 0;
+              return true;
+              // If this is weird check out explanation of pattern in ReadoutPanel
+            }
+            return false;
+          }},
+        { "FRAMES",
+          [this](const web::Animate & anim) {
+            return anim.GetFrameCount() % this->refresh_rates[this->refresh_unit];
+          }
+        }
+      };
+
+      // The current redraw checker function
+      checker_func_t do_redraw;
+
+      // A list of widget that should be redrawn when do_redraw return true
+      emp::vector<web::Widget> refresh_list;
+
+      // A function to run every frame (as fast as possible)
+      std::function<void()> simulation;
+
+      public:
+      /**
+       * Construct a shared pointer to manage ControlPanel state.
+       *
+       * @param in_id HTML ID of ConfigPanel div
+       */
+      ControlPanelInfo(const std::string & in_id="")
+      : DivInfo(in_id),
+      refresh_unit("MILLISECONDS"),
+      do_redraw(refresh_checkers.at(refresh_unit)),
+      simulation([](){ ; })  { ; }
+
+      /**
+       * Get a reference to the redraw checker function
+       * @return a redraw checker (void(const Animate &) function)
+       */
+      const checker_func_t & GetRedrawChecker() const {
+        return do_redraw;
+      }
+
+      /**
+       * Set the simulation for this control panel
+       * @param sim the function to be run every frame (as fast as possible)
+       */
+      void SetSimulation(const std::function<void()> & sim) {
+        simulation = sim;
+      }
+
+      /**
+       * Get the simulation for this control panel
+       * @return the function to be run every frame (as fast as possible)
+       */
+      const std::function<void()> & GetSimulation() const {
+        return simulation;
+      }
+
+      /**
+       * Set the refresh rate units for this control panel
+       * @param unit either "MILLISECONDS" or "FRAMES"
+       */
+      void SetUnit(const std::string & unit) {
+        refresh_unit = unit;
+        do_redraw = refresh_checkers.at(refresh_unit);
+      }
+
+      /**
+       * Set the refresh rate for this control panel
+       * @param rate the number of milliseconds or frames between refreshes
+       */
+      void SetRate(const int & rate) {
+        refresh_rates[refresh_unit] = rate;
+      }
+
+      /**
+       * Get the refresh list for this control panel
```

Comment:
Should this be private? Who needs to use this outside of the class?

Reply:
The `ControlPanel` class needs access.

Reply:
Ah yup makes sense!